<a href="https://colab.research.google.com/github/brainhack-uiowa/python-novice-inflammation-colab/blob/master/06extra1-reading-tabular-data-into-dataframes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
# setup
!wget http://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip
!unzip -n python-novice-gapminder-data.zip

--2020-06-14 02:25:53--  http://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip
Resolving swcarpentry.github.io (swcarpentry.github.io)... 185.199.111.153, 185.199.108.153, 185.199.109.153, ...
Connecting to swcarpentry.github.io (swcarpentry.github.io)|185.199.111.153|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 38471 (38K) [application/zip]
Saving to: ‘python-novice-gapminder-data.zip’

python-novice-gapmi 100%[===================>]  37.57K  --.-KB/s    in 0.01s   

2020-06-14 02:25:54 (2.55 MB/s) - ‘python-novice-gapminder-data.zip’ saved [38471/38471]

Archive:  python-novice-gapminder-data.zip
  inflating: data/gapminder_all.csv  
  inflating: data/gapminder_gdp_africa.csv  
  inflating: data/gapminder_gdp_americas.csv  
  inflating: data/gapminder_gdp_asia.csv  
  inflating: data/gapminder_gdp_europe.csv  
  inflating: data/gapminder_gdp_oceania.csv  


Original lesson and solutions can be found at [Software Carpentry](https://swcarpentry.github.io/python-novice-gapminder/07-reading-tabular/index.html).

#Reading Tabular Data into DataFrames

## Use the Pandas library to do statistics on tabular data.


- Pandas is a widely-used Python library for statistics, particularly on tabular data.
- Borrows many features from R’s dataframes.
  - A 2-dimensional table whose columns have names and potentially have different data types.
- Load it with `import pandas as pd`. The alias pd is commonly used for Pandas.
- Read a Comma Separate Values (CSV) data file with `pd.read_csv`.
  - Argument is the name of the file to be read.
  - Assign result to a variable to store the data that was read.


In [0]:
import pandas as pd

data = pd.read_csv('data/gapminder_gdp_oceania.csv')
print(data)


       country  gdpPercap_1952  ...  gdpPercap_2002  gdpPercap_2007
0    Australia     10039.59564  ...     30687.75473     34435.36744
1  New Zealand     10556.57566  ...     23189.80135     25185.00911

[2 rows x 13 columns]



- The columns in a dataframe are the observed variables, and the rows are the observations.
- Pandas uses backslash \ to show wrapped lines when output is too wide to fit the screen.


### File Not Found
Our lessons store their `data` files in a data sub-directory, which is why the path to the file is `data/gapminder_gdp_oceania.csv`. If you forget to include `data/`, or if you include it but your copy of the file is somewhere else, you will get a runtime error that ends with a line like this:

```
OSError: File b'gapminder_gdp_oceania.csv' does not exist
```

##Use `index_col` to specify that a column’s values should be used as row headings.

- Row headings are numbers (0 and 1 in this case).
- Really want to index by country.
- Pass the name of the column to `read_csv` as its `index_col` parameter to do this.



In [0]:
data = pd.read_csv('data/gapminder_gdp_oceania.csv', index_col='country')
print(data)


             gdpPercap_1952  gdpPercap_1957  ...  gdpPercap_2002  gdpPercap_2007
country                                      ...                                
Australia       10039.59564     10949.64959  ...     30687.75473     34435.36744
New Zealand     10556.57566     12247.39532  ...     23189.80135     25185.00911

[2 rows x 12 columns]


## Use `DataFrame.info` to find out more about a dataframe.

In [0]:
data.info()


<class 'pandas.core.frame.DataFrame'>
Index: 2 entries, Australia to New Zealand
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   gdpPercap_1952  2 non-null      float64
 1   gdpPercap_1957  2 non-null      float64
 2   gdpPercap_1962  2 non-null      float64
 3   gdpPercap_1967  2 non-null      float64
 4   gdpPercap_1972  2 non-null      float64
 5   gdpPercap_1977  2 non-null      float64
 6   gdpPercap_1982  2 non-null      float64
 7   gdpPercap_1987  2 non-null      float64
 8   gdpPercap_1992  2 non-null      float64
 9   gdpPercap_1997  2 non-null      float64
 10  gdpPercap_2002  2 non-null      float64
 11  gdpPercap_2007  2 non-null      float64
dtypes: float64(12)
memory usage: 208.0+ bytes



- This is a DataFrame
- Two rows named 'Australia' and 'New Zealand'
- Twelve columns, each of which has two actual 64-bit floating point values.
  - We will talk later about null values, which are used to represent missing observations.
- Uses 208 bytes of memory.


## The `DataFrame.columns` variable stores information about the dataframe’s columns.

- Note that this is data, not a method.
  - Like `math.pi`.
  - So do not use `()` to try to call it.
- Called a *member variable*, or just *member*.


In [0]:
print(data.columns)


Index(['gdpPercap_1952', 'gdpPercap_1957', 'gdpPercap_1962', 'gdpPercap_1967',
       'gdpPercap_1972', 'gdpPercap_1977', 'gdpPercap_1982', 'gdpPercap_1987',
       'gdpPercap_1992', 'gdpPercap_1997', 'gdpPercap_2002', 'gdpPercap_2007'],
      dtype='object')


## Use `DataFrame.T` to transpose a dataframe.

- Sometimes want to treat columns as rows and vice versa.
- Transpose (written `.T`) doesn’t copy the data, just changes the program’s view of it.
- Like `columns`, it is a member variable.


In [0]:
print(data.T)


country           Australia  New Zealand
gdpPercap_1952  10039.59564  10556.57566
gdpPercap_1957  10949.64959  12247.39532
gdpPercap_1962  12217.22686  13175.67800
gdpPercap_1967  14526.12465  14463.91893
gdpPercap_1972  16788.62948  16046.03728
gdpPercap_1977  18334.19751  16233.71770
gdpPercap_1982  19477.00928  17632.41040
gdpPercap_1987  21888.88903  19007.19129
gdpPercap_1992  23424.76683  18363.32494
gdpPercap_1997  26997.93657  21050.41377
gdpPercap_2002  30687.75473  23189.80135
gdpPercap_2007  34435.36744  25185.00911


##Use `DataFrame.describe` to get summary statistics about data.

`DataFrame.describe()` gets the summary statistics of only the columns that have numerical data. All other columns are ignored, unless you use the argument `include='all'`.



In [0]:
print(data.describe())


       gdpPercap_1952  gdpPercap_1957  ...  gdpPercap_2002  gdpPercap_2007
count        2.000000        2.000000  ...        2.000000        2.000000
mean     10298.085650    11598.522455  ...    26938.778040    29810.188275
std        365.560078      917.644806  ...     5301.853680     6540.991104
min      10039.595640    10949.649590  ...    23189.801350    25185.009110
25%      10168.840645    11274.086022  ...    25064.289695    27497.598692
50%      10298.085650    11598.522455  ...    26938.778040    29810.188275
75%      10427.330655    11922.958888  ...    28813.266385    32122.777857
max      10556.575660    12247.395320  ...    30687.754730    34435.367440

[8 rows x 12 columns]


- Not particularly useful with just two records, but very helpful when there are thousands.

###Exercise: Reading Other Data
Read the data in `gapminder_gdp_americas.csv` (which should be in the same directory as `gapminder_gdp_oceania.csv`) into a variable called `americas` and display its summary statistics.

###Exercise: Inspecting Data.
After reading the data for the Americas, use `help(americas.head)` and `help(americas.tail)` to find out what `DataFrame.head` and `DataFrame.tail` do.

1. What method call will display the first three rows of this data?
2. What method call will display the last three columns of this data? (Hint: you may need to change your view of the data.)





###Exercise: Reading Files in Other Directories
The data for your current project is stored in a file called `microbes.csv`, which is located in a folder called `field_data`. You are doing analysis in a notebook called `analysis.ipynb` in a sibling folder called `thesis`:
```
your_home_directory
+-- field_data/
|   +-- microbes.csv
+-- thesis/
    +-- analysis.ipynb
```
What value(s) should you pass to read_csv to read microbes.csv in analysis.ipynb?

####Solution (Brainhack provided b/c prerequisite background not covered):
We need to specify the path to the file of interest in the call to `pd.read_csv`. We first need to ‘jump’ out of the folder `thesis` using ‘../’ and then into the folder `field_data using ‘field_data/’. Then we can specify the filename `microbes.csv`. The result is as follows:

```
data_microbes = pd.read_csv('../field_data/microbes.csv')
```

### Exercise: Writing Data
As well as the `read_csv` function for reading data from a file, Pandas provides a `to_csv` function to write dataframes to files. Applying what you’ve learned about reading from files, write one of your dataframes to a file called `processed.csv`. You can use `help` to get information on how to use `to_csv`.

## Key Points


- Use the Pandas library to get basic statistics out of tabular data.
- Use `index_col` to specify that a column’s values should be used as row headings.
- Use `DataFrame.info` to find out more about a dataframe.
- The `DataFrame.columns` variable stores information about the dataframe’s columns.
- Use `DataFrame.T` to transpose a dataframe.
- Use `DataFrame.describe` to get summary statistics about data.
